actionpack/lib/action_dispatch/testing/request_encoder.rb

```diff
@@ -1,12 +1,30 @@
 # frozen_string_literal: true
 
+require "active_support/core_ext/hash/deep_transform_values"
+
 module ActionDispatch
   class RequestEncoder # :nodoc:
     class IdentityEncoder
       def content_type; end
       def accept_header; end
-      def encode_params(params); params; end
+      def encode_params(params); prevent_impossible_params(params); end
       def response_parser; -> body { body }; end
+
+      private
+
+      def prevent_impossible_params(params)
+        return params unless params.respond_to?(:deep_transform_values)
+
+        params.deep_transform_values do |v|
+          if defined?(ActiveRecord::Base) && v.is_a?(ActiveRecord::Base)
+            raise ActionDispatch::IntegrationTest::InvalidParamError.new(
+              "#{v} is an instance of an Active Record. You can't use this as a param in a test. Use basic types - strings, ints, etc - like a browser would."
```

Comment:
If there's appetite, happy to add more checks here. Another one that we use that's fairly generic is we don't allow params where `v.is_a?(Time)`; it should be a String instead.